### PR TITLE
[SPARK-30033][core] Manage shuffle IO plugins using Spark's plugin system.

### PR DESCRIPTION
**Pull request by vanzin**
SPARK-25299 is introducing a new plugin interface for shuffle IO; currently,
parts of that API provide lifecycle methods that are already covered by the
plugin API that was added in SPARK-29396.

This change makes some modifications so that:

- The driver and executor components of the shuffle plugin extend their
  respective counterparts in the generic plugin API.
- The shuffle IO plugin is managed by the same code that manages other
  generic plugins.

This simplifies and reuses similar code that exists in both implementations,
and also provides more functionality to shuffle plugins: not only do they have
more contextual information (without having to query APIs like SparkEnv) but
they also have access to other functionality in the plugin API that would
otherwise require touching internal Spark APIs.

There is a small change to the generic plugin API to avoid registering an
RPC endpoint and starting threads when not needed; plugins now must explicitly
say they want to handle RPC messages for the endpoint to be created. This is
done because the default shuffle plugin is now loaded by the plugin system,
and does not need the RPC functionality. (This API hasn't been released yet
so it's ok to make the change.)

The only downside is that initialization of the SortShuffleManager in executors
is a bit weird, because of the order in which things are initialized: the
shuffle manager is initialized by SparkEnv, and plugin initialization happens
after that. In any case, all initialization is done before any tasks are
allowed to run..

Currently, the shuffle plugin is always loaded, regardless of whether the sort
shuffle manager is being used; this was already the case in the driver, but
now is also the case in the executors. It shouldn't be hard to fix that if
needed.

Tested with existing and updated unit tests.

### CODE REVIEW

File: core/src/main/java/org/apache/spark/api/plugin/PluginRpcEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.api.plugin;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.spark.SparkContext;
```
Comment (dongjoon-hyun) [inline core/src/main/java/org/apache/spark/api/plugin/PluginRpcEndpoint.java]: Hi, @vanzin . Could you fix the `dev/lint-java`? `Collections` / `Map` / `SparkContext` seems to be unused in this interface.
```
[ERROR] src/main/java/org/apache/spark/shuffle/sort/io/LocalDiskShuffleDriverComponents.java:[24,8] (imports) UnusedImports: Unused import - org.apache.spark.SparkEnv.
38
[ERROR] src/main/java/org/apache/spark/shuffle/api/ShuffleExecutorComponents.java:[21,8] (imports) UnusedImports: Unused import - java.util.Map.
39
[ERROR] src/main/java/org/apache/spark/shuffle/api/ShuffleDriverComponents.java:[20,8] (imports) UnusedImports: Unused import - java.util.Map.
40
[ERROR] src/main/java/org/apache/spark/api/plugin/PluginRpcEndpoint.java:[20,8] (imports) UnusedImports: Unused import - java.util.Collections.
41
[ERROR] src/main/java/org/apache/spark/api/plugin/PluginRpcEndpoint.java:[21,8] (imports) UnusedImports: Unused import - java.util.Map.
42
[ERROR] src/main/java/org/apache/spark/api/plugin/PluginRpcEndpoint.java:[23,8] (imports) UnusedImports: Unused import - org.apache.spark.SparkContext.
```

**Comment by dongjoon-hyun** [comment]
It's surprising to me that the PR Builder ignores `lint-java`, but we have `GitHub Action` for that.

**Comment by tgravescs** [comment]
> The only downside is that initialization of the SortShuffleManager in executors
> is a bit weird, because of the order in which things are initialized: the
> shuffle manager is initialized by SparkEnv, and plugin initialization happens
> after that. In any case, all initialization is done before any tasks are
> allowed to run..

I haven't looked at the code yet, what exactly do you mean initialized by SparkEnv? Do you just mean that sparkEnv instantiates the shuffle manager class so its created before you possibly would want to initialize something in the plugin?  within the executor plugin it seems like having other plugins initialize first would make sense in case shuffle relied on another plugin initialization, but I supposed the reverse could be true as well.

**Comment by vanzin** [comment]
> I haven't looked at the code yet, what exactly do you mean initialized by SparkEnv?

The shuffle manager is initialized by SparkEnv (look for `shuffleManager` in the source), and the plugin context needs to be initialized after SparkEnv, currently. Kind of a chicken-and-egg thing.

Some more verbose initialization where the plugin context has multiple "init" methods (for pre- and post- SparkEnv, for example) could solve it, but then the internal state of the plugin context needs to be mutable.

Anyway, didn't look too much into making that initialization cleaner.

**Pull request closed without merging**